src/python/pants/engine/scheduler.py

```diff
@@ -327,10 +326,6 @@ def materialize_directories(self, directories_paths_and_digests):
            digest of the directories to materialize.
     :returns: Nothing or an error.
     """
-    # Ensure there isn't more than one of the same directory paths and paths do not have the same prefix.
```

Comment:
So, I think that this is a pretty reasonable check to have in place, because in concert with the "don't overwrite an existing file" case, it helps to guarantee that we don't materialize overlapping Snapshots.

If _both_ the check for existing files and this check are removed, I think that it's pretty likely for non-determinism to be introduced by having colliding snapshots materialized at the same time.

The cost of keeping this check is that in order to materialize into the same directory concurrently, callers have to use `merge_snapshots` before calling this method... that shouldn't be too, too costly, and I think remains a good defense.

Between the two checks, keeping this one feels like a better defense, because it at least protects against concurrently materializing the same file from two places. If you have that, it feels less risky to allow overwriting of files by removing the other check. But having both would be awesome.

Reply:
Added a `merge_snapshots` and re-instated this check.

Note that it only protects against concurrent modification _in one call_; multiple threads may still make materialisation requests (and ClasspathEntry's locking itself isn't currently structured in a way to avoid that, though it could be)